packages/fx-core/src/client/teamsDevPortalClient.ts

```diff
@@ -0,0 +1,947 @@
+          AppStudioError.TeamsAppPublishFailedError.message(teamsAppId, requestPath)
+        );
+      }
+    } catch (error: any) {
+      if (error instanceof SystemError) {
+        throw error;
+      } else {
+        const exception = this.wrapException(error, APP_STUDIO_API_NAMES.UPDATE_PUBLISHED_APP);
+        throw exception;
+      }
+    }
+  }
+  /**
+   * Get Stagged Teams app from tenant app catalog
+   * @param teamsAppId manifest.id, which is externalId in app catalog.
+   * @param token
+   * @returns
+   */
+  @hooks([ErrorContextMW({ source: "Teams", component: "TeamsDevPortalClient" })])
+  async getStaggedApp(
+    token: string,
+    teamsAppId: string
+  ): Promise<IPublishingAppDenition | undefined> {
+    const requester = this.createRequesterWithToken(token);
+    try {
+      const response = await RetryHandler.Retry(() =>
+        requester.get(`/api/publishing/${teamsAppId}`)
+      );
+      if (response && response.data && response.data.value && response.data.value.length > 0) {
+        const appdefinitions: IPublishingAppDenition[] = response.data.value[0].appDefinitions.map(
+          (item: any) => {
+            return {
+              lastModifiedDateTime: item.lastModifiedDateTime
+                ? new Date(item.lastModifiedDateTime)
+                : null,
+              publishingState: item.publishingState,
+              teamsAppId: item.teamsAppId,
+              displayName: item.displayName,
+            };
+          }
+        );
+        return appdefinitions[appdefinitions.length - 1];
+      } else {
+        return undefined;
+      }
+    } catch (e: any) {
+      return undefined;
+    }
+  }
+
+  @hooks([ErrorContextMW({ source: "Teams", component: "TeamsDevPortalClient" })])
+  async getUserList(token: string, teamsAppId: string): Promise<AppUser[] | undefined> {
+    const app = await this.getApp(token, teamsAppId);
+    return app.userList;
+  }
+
+  @hooks([ErrorContextMW({ source: "Teams", component: "TeamsDevPortalClient" })])
+  async checkPermission(token: string, teamsAppId: string, userObjectId: string): Promise<string> {
+    let userList;
+    try {
+      userList = await this.getUserList(token, teamsAppId);
+    } catch (error) {
+      return Constants.PERMISSIONS.noPermission;
+    }
+
+    const findUser = userList?.find((user: AppUser) => user.aadId === userObjectId);
+    if (!findUser) {
+      return Constants.PERMISSIONS.noPermission;
+    }
+
+    if (findUser.isAdministrator) {
+      return Constants.PERMISSIONS.admin;
+    } else {
+      return Constants.PERMISSIONS.operative;
+    }
+  }
+
+  @hooks([ErrorContextMW({ source: "Teams", component: "TeamsDevPortalClient" })])
+  async grantPermission(token: string, teamsAppId: string, newUser: AppUser): Promise<void> {
+    const app = await this.getApp(token, teamsAppId);
+    if (this.checkUser(app, newUser)) {
+      return;
+    }
+    app.userList?.push(newUser);
```

Comment:
`const findUser = app.userList?.findIndex((user: AppUser) => user["aadId"] === newUser.aadId);
    if (findUser && findUser >= 0) {
      return;
    }`
is removed, expected?

Reply:
checkUser already covered this, can ignore.